pilot/pkg/model/gateway.go

```diff
@@ -45,24 +56,140 @@ func MergeGateways(gateways ...Config) *MergedGateway {
+				gatewayPorts[s.Port.Number] = true
+				if isTLSServer(s) {
+					tlsServers[s.Port.Number] = []*networking.Server{s}
+				} else {
+					plaintextServers[s.Port.Number] = []*networking.Server{s}
+				}
+
+				if isHTTPServer(s) {
+					rdsRouteConfigNames[GatewayRDSRouteName(s)] = []*networking.Server{s}
+				}
 			}
 			log.Debugf("MergeGateways: gateway %q merged server %v", name, s.Hosts)
 		}
 	}
 
+	// Concatenate both sets of servers
+	for p, v := range tlsServers {
+		servers[p] = v
+	}
+	for p, v := range plaintextServers {
+		servers[p] = v
+	}
+
 	return &MergedGateway{
-		Names:   names,
-		Servers: servers,
+		Names:               names,
+		Servers:             servers,
+		RDSRouteConfigNames: rdsRouteConfigNames,
+	}
+}
+
+func isTLSServer(server *networking.Server) bool {
+	if server.Tls != nil && !ParseProtocol(server.Port.Protocol).IsHTTP() {
+		return true
 	}
+	return false
+}
+
+func isHTTPServer(server *networking.Server) bool {
+	protocol := ParseProtocol(server.Port.Protocol)
+	if protocol.IsHTTP() {
+		return true
+	}
+
+	if protocol == ProtocolHTTPS && server.Tls != nil && server.Tls.Mode != networking.Server_TLSOptions_PASSTHROUGH {
+		return true
+	}
+
+	return false
+}
+
+// GatewayRDSRouteName generates the RDS route config name for gateway's servers.
+// Unlike sidecars where the RDS route name is the listener port number, gateways have a different
+// structure for RDS.
+// HTTP servers have route name set to http.<portNumber>.
+//   Multiple HTTP servers can exist on the same port and the code will combine all of them into
+//   one single RDS payload for http.<portNumber>
+// HTTPS servers with TLS termination (i.e. envoy decoding the content, and making outbound http calls to backends)
+// will use route name https.<portnumber>.<portName>. HTTPS servers using SNI passthrough or
+// non-HTTPS servers (e.g., TCP+TLS) with SNI passthrough will be setup as opaque TCP proxies without terminating
+// the SSL connection. They would inspect the SNI header and forward to the appropriate upstream as opaque TCP.
+//
+// Within HTTPS servers terminating TLS, user could setup multiple servers in the gateway. each server could have
+// one or more hosts but have different TLS certificates. In this case, we end up having separate filter chain
+// for each server, with the filter chain match matching on the server specific TLS certs and SNI headers.
+// We have two options here: either have all filter chains use the same RDS route name (e.g. "443") and expose
+// all virtual hosts on that port to every filter chain uniformly or expose only the set of virtual hosts
+// configured under the server for those certificates. We adopt the latter approach. In other words, each
+// filter chain in the multi-filter-chain listener will have a distinct RDS route name (https.<portnumber>.portname)
+// so that when a RDS request comes in, we serve the virtual hosts and associated routes for that server.
+//
```

Comment:
Can you explain why the latter is chosen? Any advantages over the first option? I understand in the latter case that the RDS requests will be more granular, and changes in one server only requires that change to be pushed. But the same is true with HTTP, in which case the first option is chosen.